cmd/survival/main.go

```diff
@@ -0,0 +1,275 @@
+	// a casting is triggered (when `d` and `f` are invoked).
+	incantate incantate
+	spells    []string
+	point     int
+	timer     timer.Model
+	isQuit    bool
+}
+
+func main() {
+	m := model{
+		timer:     timer.NewWithInterval(timeout, time.Second),
+		orbs:      make([]string, 3, 3),
+		invoked:   make([]spell, 2, 2),
+		incantate: make(map[spell]struct{}),
+	}
+	m.cast, m.incantate, m.spells = gen(nil)
+	p := tea.NewProgram(m)
+	if _, err := p.Run(); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func (m model) Init() tea.Cmd {
+	// TODO init spells to incantate
+	// can be 2 or 3 spells
+	// can require casting or not
+	return m.timer.Init()
+}
+
+func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+	switch msg := msg.(type) {
+	case tea.KeyMsg:
+		switch msg.String() {
+		case "ctrl+c":
+			return m, tea.Quit
+		case "q", "w", "e": // handle upper case?
+			m.orbs[2] = m.orbs[1]
+			m.orbs[1] = m.orbs[0]
+			m.orbs[0] = strings.ToUpper(msg.String())
+		case "r":
+			i := invoke(m.orbs)
+			if i == Undefined {
+				break
+			}
+			// if stage only requires invoking, we can still count?
+			if !m.cast {
+				if _, ok := m.incantate[i]; ok {
+					delete(m.incantate, i)
+					m.point++
+
+					m.spells = reOrder(m.spells, spellMap[i])
+				}
+			}
+			if m.invoked[0] == i {
+				break
+			}
+			m.invoked[1] = m.invoked[0]
+			m.invoked[0] = i
+		case "d":
+			// Invoke here to prevent unnecessary reordering of spells every time a user triggers
+			// a spell casting.
+			if _, ok := m.incantate[m.invoked[0]]; ok {
+				delete(m.incantate, m.invoked[0])
+				m.point++
+				m.timer.Timeout += 2 * time.Second
+
+				m.spells = reOrder(m.spells, spellMap[m.invoked[0]])
+			}
+		case "f":
+			if _, ok := m.incantate[m.invoked[1]]; ok {
+				delete(m.incantate, m.invoked[1])
+				m.point++
+				m.timer.Timeout += 2 * time.Second
+
+				m.spells = reOrder(m.spells, spellMap[m.invoked[0]])
+			}
+		}
+	case timer.TickMsg:
+		var cmd tea.Cmd
+		m.timer, cmd = m.timer.Update(msg)
+		return m, cmd
+	case timer.TimeoutMsg:
+		m.isQuit = true
+		return m, tea.Quit
+	}
+
+	if len(m.incantate) == 0 {
+		if !m.cast {
+			m.cast, m.incantate, m.spells = gen(m.incantate)
+		} else {
+			m.cast, m.incantate, m.spells = gen(nil)
+		}
+	}
+	return m, nil
+}
+
+// TODO allocate slice
+// use strings builder
+// nice box
+// !!! center div Kappa
+func (m model) View() string {
+	var s string
+
+	if !m.cast {
+		s += "INVOKE\n\n"
+	} else {
+		s += "INVOKE AND CAST\n\n"
+	}
+
+	s += "| "
+	s += strings.Join(m.spells, " | ")
+	s += " |"
+	s += "\n--------------------------------------------------\n"
+
+	s += "| "
+	s += strings.Join(m.orbs, " | ")
+	s += " |"
+	s += "\n--------------------------------------------------\n"
+
+	s += "| Q | W | E |"
+	for _, i := range m.invoked {
+		s += " "
+		s += spellMap[i]
+		s += " |"
+	}
+	s += "\n--------------------------------------------------\n"
+
+	s += fmt.Sprintf("%d POINTS\n\n", m.point)
```

Comment:
same as above, points is not relevant in this mode

Reply:
same as above